Homework 4

LukaVukovic2 opened the pull request:
Assignment 1, Assignment 2, Assignment 3, Assignment 4, Assignment 5

- Add flexible AuthForm which handles both registration and login process
- All forms are validated with corresponding messages
- AuthRedirect component is serving as AuthGuard to prevent authorized access
- Logout functionality
- Adjuste Reviews form to use useForm hook, validation and toasts
- All forms are disabled during submitting
- Migrate to real API
- Create custom PasswordInput component to toggle password visibility
- Create tests for PasswordInput component

github-classroom[bot] committed:
Initialize assignment-3-materials

LukaVukovic2 committed:
Add new pages and sidebar navigation

Added routing for new pages using sidebar navigation and enabled dynamic show id passing via URL on show card clicks.

LukaVukovic2 committed:
Merge remote-tracking branch 'origin/assignment-3-materials' into homework-3

LukaVukovic2 committed:
Render show lists with API data

-replaced mock data with API data
-implemented flexible invalid image error
handling using components
-sorted by rating when browsing top rated shows
-added loading spinner when fetching data from API
-preserved overall website responsivity

LukaVukovic2 committed:
Create unit tests for two components

Implemented unit testing by setting up tests with goal to validate
expected input or behaviour. Tests are created for ShowCard and ReviewItem components.

LukaVukovic2 committed:
Redirection from dashboard to 'all-shows' page

LukaVukovic2 committed:
New unit tests for review input and show rendering

LukaVukovic2 committed:
Implement registration route and form component

-implement form validation (validate password confirmation) and error
handling using Yup schema builder and mutator
-implement mutators to trigger requests to the API

LukaVukovic2 committed:
Add reusable form component for login and register

Duplication of code is avoided by using AuthForm component which handles
both login and registration process by rendering correct elements and
alerts.
Upon successful login, authentication header data is stored on localStorage.

LukaVukovic2 committed:
Implement unauthorized access prevention

Added AuthRedirect component to prevent access to protected pages if not
authenticated. Also, logged in user is redirected to home page upon
visiting login/registration page.

LukaVukovic2 committed:
Implement logout functionality

Data is cleared in localStorage and deleted from cache on logout.

LukaVukovic2 committed:
Modify review form to use useForm hook

Form submission success is shown with toast component. Form controls are
disabled on all forms during submitting.

LukaVukovic2 committed:
Migrate to real API

Only authenticated users can see show lists. Old mock API data is deleted.

LukaVukovic2 committed:
Enable toggling password visibility

Created PasswordInput component which validation id and placeholder as
props. Toggling changes password visibility and shows appropriate icon.

LukaVukovic2 committed:
Write tests for new components and add layouts

PasswordInput component passed test to be correctly rendered with
correct placeholder text.
New layout shows Sidebar navigation only when logged in.

LukaVukovic2 committed:
Adjust form styles

Add corresponding icon to authentication form and rounded inputs border.
Loading spinner is added when entering submitting process.

ante-pesorda commented on homework-3/src/components/features/reviews/ReviewList/ReviewList.tsx:
```diff
@@ -0,0 +1,20 @@
+export default function ReviewList({reviews, onDeleteReview}: IReviewListProps ) {
+  return (
+    <Flex flexDirection="column">
+      {reviews.length === 0 && <p style={{margin: "10px 0"}}>No reviews yet</p>}
+      {reviews.map((review, index) => {
```
Možemo li ovo postavit kao if-else blok? Da se osiguramo da se nikad neće renderat oboje

ante-pesorda commented on homework-4/src/app/(shows)/logout/page.tsx:
treba li nam page za logout?

ante-pesorda commented on homework-4/src/components/features/auth/LoginForm/LoginForm.tsx:
Ne bih koristio yup paket naše forme. U ovoj zadaći je ideja da koristimo useForm hook

ante-pesorda commented on homework-4/src/components/features/shows/ReviewForm/ReviewForm.tsx:
```diff
@@ -0,0 +1,182 @@
+  rating: number;
+}
+
+const schema = yup
+  .object({
+    comment: yup.string().required("Comment is required"),
+    rating: yup
+      .number()
+      .required("Rating is required")
+      .min(1, "Rating must be at least 1 star")
+      .max(5, "Rating must be at most 5 stars"),
+  })
+  .required();
```
I ovdje - useForm hook, na kojem onda registriramo inpute i hendlamo errore i stateove

ante-pesorda commented on homework-4/src/components/features/shows/ShowDetailsSection/ShowDetailsSection.tsx:
```diff
@@ -0,0 +1,32 @@
+    return <div>No show available</div>;
+  }
+
+  if (isLoading || !data) {
+    return <LoadingSpinner />;
+  }
```
Kad iz nekog razloga ne dobijemo podatke i dalje će se nastavit vrtit spinner, što je loš UX.
Trebalo bi returnati ili neku poruku ili empty state komponentu. Odnosno, treba li nam ovdje provjera za `!data`

LukaVukovic2 committed:
Code refactor

-use useForm hook validation instead of Yup
-add logout function which handles logout process
-prevent infinite loading spinner effect on error

ante-pesorda reviewed (commented):
Još bi jedino postavio samo jedan localStorage item kod autentifikacije, a ne tri kao sad

LukaVukovic2 committed:
Change localStorage authentication item

Combine localStorage auth items into one "user" item

ante-pesorda reviewed (approved):
🔥
